File: TinyScript/JAVA/src/translator/symbol/SymbolTable.java
package translator.symbol;

import lexer.Token;
import lexer.TokenType;

import java.util.ArrayList;

public class SymbolTable {
    public SymbolTable parent = null;
    // ArrayList 可以自动扩容
    private ArrayList<SymbolTable> children;
    private ArrayList<Symbol> symbols;

    /**
     * 为三地址码临时创建的变量分配记录 id 的
     * 如：var a = 1+2*3
     * p0=2*3，这时 temIndex 就给 p0 分配个id
     */
    private int temIndex = 0;

    /**
     * 为创建的变量记录 offset 索引，如：
     * var a = 1;
     * var b = 1;
     * var c = 3;
     * 此时 offsetInde 就记录他们的 offset index，写入内存时就参照这个index
     */
    private int offsetIndex = 0;

    // 当前节点距离根结点的距离
    private int level = 0;

    public SymbolTable() {
        this.children = new ArrayList<>();
        this.symbols = new ArrayList<>();
    }

    // 为当前结点添加子结点
    public void addSymbol(Symbol symbol) {
        this.symbols.add(symbol);
        symbol.parent = this;
    }

    public boolean exists(Token lexeme) {
        var _symbol = this.symbols.stream()
                .filter(x -> x.lexeme.getValue().equals(lexeme.getValue()))
                .findFirst();
        if (!_symbol.isEmpty()) {
            return true;
        }

        if (this.parent != null) {
            return this.parent.exists(lexeme);
        }

        return false;
    }

    public Symbol cloneFromSymbolTree(Token lexeme, int layoutOffset) {
        var _symbol = this.symbols.stream()
                .filter(x -> x.lexeme.getValue().equals(lexeme.getValue()))
                .findFirst();

        if (!_symbol.isEmpty()) {
            var symbol = _symbol.get().copy();
            symbol.setLayerOffset(layoutOffset);
            return symbol;
        }

        if (this.parent != null) {
            return this.parent.cloneFromSymbolTree(lexeme, layoutOffset + 1);
        }

        return null;
    }

    public Symbol createSymbolByLexeme(Token lexeme) {
        Symbol symbol = null;

        // 如果是个标量
        if (lexeme.isScalar()) {
            symbol = Symbol.createImmediateSymbol(lexeme);
        } else {
            // 如果是变量之类的要先克隆，以便记录 layerOffset 层数
            symbol = cloneFromSymbolTree(lexeme, 0);

            if (symbol == null) {
                symbol = Symbol.createAddressSymbol(lexeme, this.offsetIndex++);
            }
        }
        this.symbols.add(symbol);
        return symbol;
    }

    // 创建临时变量
    public Symbol createVariable() {
        var lexeme = new Token(TokenType.VARIABLE, "p" + this.temIndex++);
        var symbol = Symbol.createAddressSymbol(lexeme, this.offsetIndex++);
        this.addSymbol(symbol);
        return symbol;
    }

    // 为它增加一个子符号表
    public void addChild(SymbolTable child) {
        child.parent = this;
        child.level = this.level + 1;
        this.children.add(child);
    }

    // 查看当前符号表占据了多大的空间
    public int localSize() {
        return this.offsetIndex;
    }

    public ArrayList<SymbolTable> getChildren() {
        return this.children;
    }

    public void createLabel(String label, Token lexeme) {
        var labelSymbol = Symbol.createLabelSymbol(label, lexeme);
        this.addSymbol(labelSymbol);
    }
}

File: TinyScript/JS/src/translator/symbol/SymbolTable.js
const Token = require('../../lexer/Token')
const TokenType = require('../../lexer/TokenType')
const Symbol = require('./Symbol')

class SymbolTable {
  constructor() {
    this.parent = null
    this.symbols = []
    this.children = []

    // 用来生成分配常量的三地址码临时变量的 ID 的
    this.tempIndex = 0
    this.offsetIndex = 0
    this.level = 0
  }

  addSymbol(symbol) {
    this.symbols.push(symbol)
    symbol.setParent(this)
  }

  addChild(child) {
    child.parent = this
    child.level = this.level + 1
    this.children.push(child)
  }

  getSymbols() {
    return this.symbols
  }

  getChildren() {
    return this.children
  }


  // 在一个符号表里创造一个标签
  createLabel(label, lexeme) {
    const labelSymbol = Symbol.createLabelSymbol(label, lexeme)
    this.addSymbol(labelSymbol)
  }

  // 创建临时变量
  createVariable() {
    const lexeme = new Token(TokenType.VARIABLE, "p" + this.tempIndex++)
    const symbol = Symbol.createAddressSymbol(lexeme, this.offsetIndex++)
    this.addSymbol(symbol)
    return symbol
  }

  // 判断当前符号表里是否存在 指定 的符号
  exists(lexeme) {
    const symbol = this.symbols.find(x => x.lexeme.getValue() == lexeme.getValue())

    if (symbol) {
      return true
    }

    if (this.parent != null) {
      return this.parent.exists(lexeme)
    }

    return false
  }

  // 把（父层）变量 symbol 拷贝到当前层级，并重新维护它的 layerOffset
  cloneFromSymbolTree(lexeme, layerOffset) {
    let symbol = this.symbols
      .find(x => x.lexeme.getValue() == lexeme.getValue())

    if (symbol) {
      symbol = symbol.copy()
      symbol.setLayerOffset(layerOffset)
      return symbol
    }

    if (this.parent != null) {
      return this.parent.cloneFromSymbolTree(lexeme, layerOffset + 1)
    }

    return null
  }

  // 通过 lexeme 创建一个 Symbol
  createSymbolByLexeme(lexeme) {
    let symbol = null

    if (lexeme.isScalar()) {
      symbol = Symbol.createImmediateSymbol(lexeme)
    } else {
      symbol = this.cloneFromSymbolTree(lexeme, 0)
      if (symbol == null) {
        symbol = Symbol.createAddressSymbol(lexeme, this.offsetIndex++)
      }
    }

    this.addSymbol(symbol)
    return symbol
  }

  createVariable() {
    var lexeme = new Token(TokenType.VARIABLE, "p" + this.tempIndex++)
    var symbol = Symbol.createAddressSymbol(lexeme, this.offsetIndex++)
    this.addSymbol(symbol)
    return symbol
  }

  // 当前的栈有几个占地址的元素（即我们 push和add 进去的元素）
  localSize() {
    return this.offsetIndex
  }
}

module.exports = SymbolTable